proc-macros/src/new/render_server.rs

```diff
@@ -0,0 +1,234 @@
+			let rpc_method_name = self.rpc_identifier(&method.name);
+			// `parsing` is the code associated with parsing structure from the
+			// provided `RpcParams` object.
+			// `params_seq` is the comma-delimited sequence of parametsrs.
+			let (parsing, params_seq) = self.render_params_decoding(&method.params);
+
+			if method.signature.sig.asyncness.is_some() {
+				quote! {
+					rpc.register_async_method(#rpc_method_name, |params, context| {
+						let owned_params = params.owned();
+						let fut = async move {
+							let params = owned_params.borrowed();
+							#parsing
+							Ok(context.as_ref().#rust_method_name(#params_seq).await)
+						};
+						Box::pin(fut)
+					})?;
+				}
+			} else {
+				quote! {
+					rpc.register_method(#rpc_method_name, |params, context| {
+						#parsing
+						Ok(context.#rust_method_name(#params_seq))
+					})?;
+				}
+			}
+		});
+
+		let subscriptions = self.subscriptions.iter().map(|sub| {
+			// Rust method to invoke (e.g. `self.<foo>(...)`).
+			let rust_method_name = &sub.signature.sig.ident;
+			// Name of the RPC method to subscribe (e.g. `foo_sub`).
+			let rpc_sub_name = self.rpc_identifier(&sub.name);
+			// Name of the RPC method to unsubscribe (e.g. `foo_sub`).
+			let rpc_unsub_name = self.rpc_identifier(&sub.unsub_method);
+			// `parsing` is the code associated with parsing structure from the
+			// provided `RpcParams` object.
+			// `params_seq` is the comma-delimited sequence of parametsrs.
+			let (parsing, params_seq) = self.render_params_decoding(&sub.params);
+
+			quote! {
+				rpc.register_subscription(#rpc_sub_name, #rpc_unsub_name, |params, sink, context| {
+					#parsing
+					Ok(context.as_ref().#rust_method_name(sink, #params_seq))
+				})?;
+			}
+		});
+
+		let doc_comment = "Collects all the methods and subscriptions defined in the trait \
+								and adds them into a single `RpcModule`.";
+
+		Ok(quote! {
+			#[doc = #doc_comment]
+			fn into_rpc(self) -> Result<#rpc_module<Self>, #jrps_error> {
+				let mut rpc = #rpc_module::new(self);
+
+				#(#methods)*
+				#(#subscriptions)*
+
+				Ok(rpc)
+			}
+		})
+	}
+
+	fn render_params_decoding(&self, params: &[(syn::PatIdent, syn::Type)]) -> (TokenStream2, TokenStream2) {
+		if params.is_empty() {
+			return (TokenStream2::default(), TokenStream2::default());
+		}
+
+		let jrps_call_error = self.jrps_server_item(quote! { error::CallError });
+		let serde_json = self.jrps_server_item(quote! { __reexports::serde_json });
+
+		// Parameters encoded as a tuple (to be parsed from array).
+		let (params_fields_seq, params_types_seq): (Vec<_>, Vec<_>) = params.iter().cloned().unzip();
+		let params_types = quote! { (#(#params_types_seq),*) };
+		let params_fields = quote! { (#(#params_fields_seq),*) };
+
+		// Code to decode sequence of parameters from a JSON array.
+		let decode_array = {
+			let decode_fields = params.iter().enumerate().map(|(id, (name, ty))| {
+				if is_option(ty) {
```

Comment:
am I reading this correct; this actually provides `None` if the param is missing in the request? If so, neat.

Reply:
Yep, I thought it will be strictly incorrect to explicitly require optional arguments. Missing optional arguments can either be provided as `null` or not be included at all.